File: src/mod11_gringotts_concurrency/lesson11_thread_pools/java/Main.java
package mod11_gringotts_concurrency.lesson11_thread_pools.java;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class Main {
    public static void main(String[] args) {

        BankAccount ollivander = new BankAccount("Ollivander's account", 15_000);
        BankAccount hagrid = new BankAccount("Hadrid's account", 500_000);

        //newFixedThreadPool()
//        ExecutorService executor = Executors.newFixedThreadPool(2);
//        executor.submit(new Thread1(hagrid, ollivander));
//        executor.submit(new RunnableForThread2(hagrid, ollivander));
//        executor.shutdown();

        //newCachedThreadPool()
//        ExecutorService executor = Executors.newCachedThreadPool();
//        executor.submit(() -> hagrid.transferFunds(ollivander, 5_000));
//        executor.submit(() -> ollivander.transferFunds(hagrid, 3_000));
//        executor.shutdown();

        //newScheduledThreadPool()
//        ScheduledExecutorService executor = Executors.newScheduledThreadPool(2);
//        executor.schedule(new Thread1(hagrid, ollivander),10, TimeUnit.SECONDS);
//        executor.schedule(new RunnableForThread2(hagrid, ollivander),15, TimeUnit.SECONDS);
//        executor.shutdown();

        //newScheduledThreadPool()
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.submit(new Thread1(hagrid, ollivander));
        executor.submit(new RunnableForThread2(hagrid, ollivander));
        executor.shutdown();

        try {
            executor.awaitTermination(1, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        System.out.println("Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
                ollivander.getAccountName() + ":" + ollivander.getBalance());
    }
}

File: src/mod11_gringotts_concurrency/lesson11_thread_pools/kotlin/Main.kt
package mod11_gringotts_concurrency.lesson11_thread_pools.kotlin

import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

fun main() {

    val hagridAccount = BankAccount("Hadrid's account", 500_000)
    val ollivanderAccount = BankAccount("Ollivander's account", 15_000)
//    val executorService = Executors.newFixedThreadPool(2)
//    val executorService = Executors.newCachedThreadPool()
//    val executorService = Executors.newScheduledThreadPool(2)
//    executorService.schedule({
//        repeat(100) {
//            hagridAccount.transferFunds(ollivanderAccount, 10)
//        }
//    }, 5, TimeUnit.SECONDS)
//    executorService.schedule({
//        repeat(100) {
//            hagridAccount.transferFunds(ollivanderAccount, 10)
//        }
//    },10, TimeUnit.SECONDS)

    val executorService = Executors.newSingleThreadExecutor()

    executorService.submit {
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
    }
    executorService.submit {
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
    }

    executorService.shutdown()
    executorService.awaitTermination(1, TimeUnit.MINUTES)

    println(
        "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
                ollivanderAccount.accountName + ":" + ollivanderAccount.balance
    )


}